rfcs/rfc0010.md

```diff
@@ -0,0 +1,157 @@
+an array using two-target foreach, we might write this:
+
+```perl
+for my ($i, $value) (%array[ keys @array ]) {
+  say "$i == $value";
+}
+```
+
+This is tolerable, but a bit verbose.  If we bury our target array deep in a
+structure, we get this:
+
+```perl
+for my ($i, $value) ($alpha->{beta}->[0]->%[ keys $alpha->{beta}->[0]->@* ]) {
+  say "$i == $value";
+}
+```
+
+This is pretty bad.
+
+With `kv`, we write this:
+
+```perl
+for my ($i, $value) (kv $alpha->{beta}->[0]->@*) {
+  say "$i == $value";
+}
+```
+
+This is probably about as simple as this can get without some significant new
+addition to the language.
+
+## Specification
+
+Like `keys` or `values`, `kv` can be called with a hash expression or array
+expression.  Other types of arguments are compile time errors.
+
+In scalar context, `kv` returns the number of entries in its argument, just
+like `keys` or `values`.
+
+In list context, `kv %h` is identical in behavior to `%h` in list context.
+
+In list context, `kv @arr` returns a list twice the size of the array, zipping
+together the keys and values.  To match existing behaviors, the keys are
+mutable copies and the values are aliases.  This requires the following work:
+
+```
+  DB<1> @arr = (1..10)
+
+  DB<2> x \@arr
+0  ARRAY(0x7f9784b09ed8)
+   0  1
+   1  2
+   2  3
+   3  4
+   4  5
+   5  6
+   6  7
+   7  8
+   8  9
+   9  10
+
+  DB<3> $_ ++ for kv @arr
+
+  DB<4> x \@arr
+0  ARRAY(0x7f9784b09ed8)
+   0  2
+   1  3
+   2  4
+   3  5
+   4  6
+   5  7
+   6  8
+   7  9
+   8  10
+   9  11
+```
+
+## Backwards Compatibility
+
+There should be no significant backwards compatibility concerns.  `kv` will be
+imported only when requested.  Static analysis tools may need to be updated.
+
+For polyfill in older perls, I don't believe a pure-perl implementation of `kv`
+can be provided while maintaining its aliasing behavior.  This may not be true.
+
+## Security Implications
+
+Nothing specific predicted.
+
+## Examples
+
+(See the examples under **Rationale**.)
+
+I expect that docs for `keys` and `values` will be updated to reference `kv` as
+well, and we'll add a note about it to the documentation on `for` and possibly
+pair slices.
+
+When n-at-a-time foreach is no longer experimental, we should refer to these
+two features together as an alternative to `each` in its documentation.
+
+## Prototype Implementation
+
+None.
+
+## Future Scope
+
+I believe this will be complete as is.
+
+## Rejected Ideas
+
+One suggestion was that `kv LIST` should zip together the list with would-be
+indexes.  I believe this is a complication, as it multiplies the *kind* of
+builtins surrounding container inspection.  My take is that `keys` and `values`
+don't work on lists, so this also should not.
+
+## Open Issues
+
+* How will this be requested?  The answer is likely: If the "builtin" namespace
+  exists when this is ready, we use that.  If not, it goes behind a `feature`
```

Comment:
Actually I disagree. `builtin` is for things that are purely functions, but as `kv` is like `keys` and `values` in that they are really basic syntax operators that operate directly on an array or hash term and not just "and ol' list", they should not be in that space.

My suggested "enumerate" or whatever it might be called, could be a builtin function. But this `kv` really is a real operator, not just a function.